### PR TITLE
Write forecast to database

### PR DESCRIPTION
**Pull request by simlmx**
_I recommend reviewing each commit separately._

* Write to the database
* Use DatabaseConnection
* Some mypy fixes in the github workflow

## Checklist:

- [x] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/nowcasting/blob/main/coding_style.md)
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: pv_site_production/app.py [hunk 1]
```diff
@@ -86,10 +93,21 @@ def main(
...
+        # TODO Make `insert_forecast_values` support (expect?) datetimes.
+        # Currently `insert_forecast_values` expects datetimes as string with ISO formatting.
+        results_df["target_datetime_utc"] = results_df["target_datetime_utc"].dt.strftime(
+            "%Y-%m-%dT%H:%M:%SZ"
+        )
+
+        with database_connection.get_session() as session:  # type: ignore
+            insert_forecast_values(session, results_df)
+    else:
+        # When we don't write to the DB, we print to stdout instead.
```
Comment (peterdudfield) [inline pv_site_production/app.py:108]: nice

File: tests/conftest.py [hunk 2]
```diff
@@ -17,56 +18,47 @@
...
+        yield database_connection
 
-    You can use `with Session() as session:`
-    """
-    return sessionmaker(bind=engine)
+        engine.dispose()
 
 
 @pytest.fixture()
-def db_session(Session):
```
Comment (peterdudfield) [inline tests/conftest.py:51]: you can even make a sessions fixture, then most other fixtures just use that, so they dont have to worry about not closing the database connection e.t.c

File: tests/test_app.py [hunk 3]
```diff
@@ -12,10 +12,12 @@
 
 
 @pytest.mark.parametrize("config_file", CONFIG_FIXTURES)
-def test_app(config_file: pathlib.Path):
+@pytest.mark.parametrize("write_to_db", [True, False])
```
Comment (peterdudfield) [inline tests/test_app.py:15]: nice

File: tests/test_app.py [hunk 4]
```diff
@@ -12,10 +12,12 @@
...
+def test_app(config_file: pathlib.Path, write_to_db: bool):
     runner = CliRunner()
     result = runner.invoke(
         main,
-        ["--config", str(config_file), "--date", "2022-1-1-11-50"],
+        ["--config", str(config_file), "--date", "2022-1-1-11-50"]
+        + (["--write-to-db"] if write_to_db else []),
     )
     assert result.exit_code == 0, traceback.print_exception(result.exception)
```
Comment (peterdudfield) [inline tests/test_app.py:23]: would be good to check that thing have been written to the database in this test

File: tests/conftest.py (hunk 2 above)
Comment (simlmx) [reply tests/conftest.py:51]: Isn't it what this fixture still is?

File: tests/test_app.py (hunk 4 above)
Comment (simlmx) [reply tests/test_app.py:23]: Yes I agree. I didn't do it out of lazyness 😅 . Will add.

File: tests/conftest.py [hunk 5]
```diff
@@ -118,35 +110,6 @@ def db_data(database_connection):
 
         session.commit()
 
-        # Forecast
```
Comment (simlmx) [inline tests/conftest.py:129]: Those are not used by any test.

**Review by devsjc — APPROVED** [review]
Nicely done @simlmx!

File: tests/conftest.py (hunk 2 above)
Comment (peterdudfield) [reply tests/conftest.py:51]: oh sorry

**Pull request merged**